File: bffs/mobile-bff/app/clients/user_client.py
import httpx
from app.config import USER_SERVICE_URL
from app.utils.circuit_breaker import create_breaker


register_user_breaker = create_breaker("user.register")
login_user_breaker = create_breaker("user.login")
get_user_breaker = create_breaker("user.getUser")


async def _register_user(payload: dict):
    async with httpx.AsyncClient(timeout=10) as client:
        response = await client.post(f"{USER_SERVICE_URL}/api/users/register", json=payload)
        response.raise_for_status()
        return response.json()


async def _login_user(payload: dict):
    async with httpx.AsyncClient(timeout=10) as client:
        response = await client.post(f"{USER_SERVICE_URL}/api/users/login", json=payload)
        response.raise_for_status()
        return response.json()


async def _get_user(user_id: str, authorization: str | None = None):
    headers = {}

    if authorization:
        headers["Authorization"] = authorization

    async with httpx.AsyncClient(timeout=10) as client:
        response = await client.get(
            f"{USER_SERVICE_URL}/api/users/{user_id}",
            headers=headers
        )
        response.raise_for_status()
        return response.json()


async def register_user(payload: dict):
    return await register_user_breaker.call(_register_user, payload)


async def login_user(payload: dict):
    return await login_user_breaker.call(_login_user, payload)


async def get_user(user_id: str, authorization: str | None = None):
    return await get_user_breaker.call(_get_user, user_id, authorization)

File: bffs/web-bff/src/clients/userClient.js
const env = require("../config/env");
const httpClient = require("./httpClient");
const { createBreaker } = require("../utils/circuitBreaker");

const registerBreaker = createBreaker("user.register", async (body) => {
  const response = await httpClient.post(`${env.userServiceUrl}/api/users/register`, body);
  return response.data;
});

const loginBreaker = createBreaker("user.login", async (body) => {
  const response = await httpClient.post(`${env.userServiceUrl}/api/users/login`, body);
  return response.data;
});

const getUsersBreaker = createBreaker(
  "user.getUsers",
  async (headers) => {
    const response = await httpClient.get(`${env.userServiceUrl}/api/users`, {
      headers
    });
    return response.data;
  },
  () => []
);

const getUserBreaker = createBreaker("user.getUser", async (id, headers) => {
  const response = await httpClient.get(`${env.userServiceUrl}/api/users/${id}`, {
    headers
  });
  return response.data;
});

const updateUserBreaker = createBreaker("user.updateUser", async (id, body, headers) => {
  const response = await httpClient.put(`${env.userServiceUrl}/api/users/${id}`, body, {
    headers
  });
  return response.data;
});

const deleteUserBreaker = createBreaker("user.deleteUser", async (id, headers) => {
  const response = await httpClient.delete(`${env.userServiceUrl}/api/users/${id}`, {
    headers
  });
  return response.data;
});

const userClient = {
  register: (body) => registerBreaker.fire(body),
  login: (body) => loginBreaker.fire(body),
  getUsers: (headers) => getUsersBreaker.fire(headers),
  getUser: (id, headers) => getUserBreaker.fire(id, headers),
  updateUser: (id, body, headers) => updateUserBreaker.fire(id, body, headers),
  deleteUser: (id, headers) => deleteUserBreaker.fire(id, headers)
};

module.exports = userClient;